Blazor Hot Reload 7.0 improvements (#27574)
dotnet/AspNetCore.Docs: aspnetcore/test/hot-reload.md

## Blazor WebAssembly

::: moniker range=">= aspnetcore-7.0"

Blazor WebAssembly Hot Reload supports the following code changes:

* New types.
* Nested classes.
* Most changes to method bodies, such as adding, removing, and editing variables, expressions, and statements.
* Changes to the bodies of [lambda expressions](/dotnet/csharp/language-reference/operators/lambda-expressions) and [local functions](/dotnet/csharp/programming-guide/classes-and-structs/local-functions).
* Adding static and instance methods to existing types.
* Adding static fields and methods to existing types.
* Adding static lambdas to existing methods.
* Adding lambdas that capture `this` to existing methods that already captured `this` previously.

> Note that when an attribute is removed that previously set the value of a component parameter, the component parameter is set back to its default value.

Reviewer: ```suggestion
Note that when an attribute is removed that previously set the value of a component parameter, the component is disposed and re-initialized to set the removed parameter back to its default value.
```

## Blazor WebAssembly

Blazor WebAssembly Hot Reload support has the following conditions:
::: moniker range=">= aspnetcore-7.0"

> * Hot Reload reacts to most changes to method bodies, such as adding, removing, and editing variables, expressions, and statements.

Reviewer 2: The first two bullet points in this list are still valid. The list you added below is for the edits that we *added* support for in .NET 7.

* New types.
* Nested classes.
* Most changes to method bodies, such as adding, removing, and editing variables, expressions, and statements.
* Changes to the bodies of [lambda expressions](/dotnet/csharp/language-reference/operators/lambda-expressions) and [local functions](/dotnet/csharp/programming-guide/classes-and-structs/local-functions).
* Adding static and instance methods to existing types.
* Adding static fields and methods to existing types.
* Adding static lambdas to existing methods.
* Adding lambdas that capture `this` to existing methods that already captured `this` previously.

> Blazor WebAssembly and Blazor Server Hot Reload have support for when components reset their parameters to their default values when a value is removed.

Reviewer 2: ```suggestion
Note that when an attribute is removed that previously set the value of a component parameter, the component parameter is set back to its default value.
```

@MackinnonBuck What do we use as the "default" value? Do we just do `default<T>` or do we look at any default values for the property?

Reviewer: @danroth27 What we do is if we detect that a component had one or more parameters removed in a hot reload edit, we completely destroy and re-create the component. This allows any custom component initialization logic to set missing parameters to their intended "default" values.

Reviewer: So in other words, the "default" value is the value that the parameter would have if it were unspecified.

The following code changes aren't supported for Blazor WebAssembly apps:

> * Adding local functions.

Reviewer 3: ```suggestion
```

* Adding a new [`await` operator](/dotnet/csharp/language-reference/operators/await) or [`yield` keyword](/dotnet/csharp/language-reference/keywords/yield) expression.
* Changing the names of method parameters.

> * Changes outside of method bodies.

Reviewer 3: ```suggestion
* Adding new non-`static` fields, events or properties 
```
